### ISSUE TITLE
New Route children API does not forward RouterContext props

### ISSUE THREAD

**Issue by dylinmaust** [ReactTraining/react-router#6960]
## Version

react-router-dom 5.1.2

## Test Case

https://codesandbox.io/embed/loving-mcnulty-ygk9u

## Steps to reproduce
1. Navigate to Topics route

## Expected Behavior
Topics function renders and displays nested routes using the match param provided via Router context.

## Actual Behavior
Topics function throws an error because match is undefined. Indeed, no props are forwarded to the Topics function.

The error reads: `Cannot read property 'url' of undefined`

**Comment by leoyli** [comment]
I think this is expected. Here is the pattern shifts.

Previously it was using the render prop pattern, or pass component with `component` prop to inject the routing contexts. Now to make it more declarative, we can use children + hooks. In the code sandbox, the `Topic` component need to call `useRouteMatch` hook to get the `match` object.

**Comment by dylinmaust** [comment]
Interesting. The CodeSandbox code above is taken directly from https://reacttraining.com/react-router/web/guides/quick-start where `match` is provided to the child component without any use of hooks. Additionally, I thought the main purpose of `useRouteMatch` was to replace floating `<Route /> components that were used solely for getting `match` data.

Additionally, it looks like React Router should still be passing the RouterContext props to Route's `children`, unless I'm misunderstanding the source: https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Route.js#L56

**Comment by MeiKatz** [comment]
@dylinmaust `<Route />` passes the routing props to `children` if (and only if) `children` is a function. Any documentation that say something else is wrong and needs to be fixed.

**Comment by dylinmaust** [comment]
@leoyli Good call! I didn't realize you could just use `useRouteMatch` without passing params to get the current match. It seems that this is a documentation issue rather than a bug within React Router.

**Comment by leoyli** [comment]
There are currently 3 patterns:

1. render props, children as function:

```js
<Route render={(props) => <Child {...props} />} />
```

```js
<Route>(props) => <Child {...props} /></Route>
```

They are, in general, the same beings.

2. component

```js
<Route component={Child} />
```

Please note, it is `Child`, not `<Child />`.

3. element

```js
<Route>
  <Child /> {/* you need to call hook inside the children instead. */}
</Route>
```

So basically, the 1st and 2nd are the old patterns we are used to. The third one is special in the way it now can be passed as an element, and since we did not provide any props, we need to "inject" the contexts from the hooks.

**Comment by MeiKatz** [comment]
The 3rd one is the one that will stay in the future. `render` and `component` may be dropped in the future, because you can do both also with `children`.

**Comment by MeiKatz** [comment]
@leoyli You're right: we should add a detection for lazy loaded components so we could do something like this:
```jsx
<Route path="/foo">
  {React.lazy(() => import("./foo"))}
</Route>
```

**Comment by timdorr** [comment]
Fixed by #6952. Might just need to be published.
